las/credentials.py

```diff
@@ -1,68 +1,108 @@
 import configparser
-
 import os
-from os.path import expanduser, exists
-from typing import Tuple, Optional
+import time
+from os.path import exists, expanduser
+from typing import Optional, Tuple, List
+
+import requests
+from requests.auth import HTTPBasicAuth
 
 
 class MissingCredentials(Exception):
     pass
 
 
 class Credentials:
-    """Used to fetch and store credentials. One of 4 conditions must be met to successfully create credentials.
+    """Used to fetch and store credentials. One of 3 conditions must be met to successfully create credentials.
 
     1. credentials_path is provided
-    2. access_key_id, secret_access_key and api_key is provided
-    3. credentials is located in default path ~/.lucidtech/credentials.cfg
-    4. the following environment variables are present:
-        - LAS_ACCESS_KEY_ID
-        - LAS_SECRET_ACCESS_KEY
+    2. credentials is located in default path ~/.lucidtech/credentials.cfg
+    3. the following environment variables are present:
+        - LAS_CLIENT_ID
+        - LAS_CLIENT_SECRET
         - LAS_API_KEY
+        - LAS_AUTH_ENDPOINT
+        - LAS_API_ENDPOINT
 
     :param credentials_path: Path to credentials file
     :type credentials_path: str
-    :param access_key_id: Access Key Id
-    :type access_key_id: str
-    :param secret_access_key: Secret Access Key
-    :type secret_access_key: str
+    :param client_id: Client Id
+    :type client_id: str
+    :param client_secret: Client Secret
+    :type client_secret: str
     :param api_key: API key
     :type api_key: str
+    :param auth_endpoint: Authorization endpoint
+    :type auth_endpoint: str
 
     """
-    def __init__(self, credentials_path=None, access_key_id=None, secret_access_key=None, api_key=None):
-        if not all([access_key_id, secret_access_key, api_key]):
-            if not credentials_path:
-                credentials_path = expanduser('~/.lucidtech/credentials.cfg')
+    def __init__(self, credentials_path=None):
+        self._token = (None, None)
+        creds = list(self._read_from_file(credentials_path))
+        creds_from_environ = self._read_from_environ()
 
-            if exists(credentials_path):
-                access_key_id, secret_access_key, api_key = self._read_from_file(credentials_path)
-            else:
-                access_key_id, secret_access_key, api_key = self._read_from_environ()
+        for idx, value in enumerate(creds):
+            if creds_from_environ[idx] is not None:
+                creds[idx] = creds_from_environ[idx]
```

Comment:
I don't think ENV credentials should override credentials provided directly, it will be confusing for users that explicitly provided credentials are not used. The way we default initialize creds in the method param is problematic when determining if we should look at ENV or not. I suggest changing the constructor.